pymongo/mongo_client.py

```diff
@@ -2086,6 +2089,7 @@ class _MongoClientErrorHandler(object):
         "completed_handshake",
         "service_id",
         "handled",
+        "sock_info",
```

Comment:
I'm a little wary of storing the sock_info in more places since we've had odd garbage collection issues in the past (https://jira.mongodb.org/browse/PYTHON-2751 for one). 

What if we keep the same contribute_socket behavior as before but add a completed_handshake=True arg then in pool.connect:
```python
            if self.handshake:
                sock_info.hello()
                self.is_writable = sock_info.is_writable
            if handler:
                 handler.contribute_socket(sock_info, complete_handshake=False)
```

Reply:
Done